river/PointerMapping.zig

```diff
@@ -14,13 +14,47 @@
 // You should have received a copy of the GNU General Public License
 // along with this program. If not, see <https://www.gnu.org/licenses/>.
 
+const Self = @This();
+
 const wlr = @import("wlroots");
 
+const util = @import("util.zig");
+
 pub const Action = enum {
     move,
     resize,
+    command,
 };
 
 event_code: u32,
 modifiers: wlr.Keyboard.ModifierMask,
 action: Action,
+command_args: []const [:0]const u8,
+
+pub fn init(
+    event_code: u32,
+    modifiers: wlr.Keyboard.ModifierMask,
+    action: Action,
+    command_args: []const [:0]const u8,
+) !Self {
+    const args_len = if (action == Action.command) command_args.len else 0;
+    const owned_args = try util.gpa.alloc([:0]u8, args_len);
```

Comment:
Even for static length arrays, it's probably a good idea to use `ArrayListUnmanaged`, since it's API is a bit more handy.

Roughly like this:
```zig
var owned_args: std.ArrayListUnmanaged([:0]u8) = .{};
try owned_args.ensureTotalCapacity(util.gpa, command_args.len);
errdefer {
    for (owned_args.items) |arg| util.gpa.free(arg);
    owned.deinit(util.gpa);
}
for (command_args) |arg| {
    const owned = try util.gpa.dupeZ(arg);
    owned_args.appendAssumeCapacity(owned);
}
```

It might even make sense to put these allocations into an arena for easier cleanup, maybe (Just make sure to not grow the array list after the initial setup).

Reply:
I'm struggling with the use of an arena. Attempt:

```zig
        ActionType.command => blk: {
            var arena = std.heap.ArenaAllocator.init(util.gpa);
            defer arena.deinit();

            var owned_args = try std.ArrayListUnmanaged([:0]const u8).initCapacity(arena.allocator(), command_args.len);

            for (command_args) |arg| {
                const owned = try util.gpa.dupeZ(u8, arg);
                owned_args.appendAssumeCapacity(owned);
            }

            break :blk Action{ .command = owned_args.toOwnedSlice(util.gpa) };
        },
```

Even though the strings and the slice are allocated using gpa the arena always frees everything.

Reply:
The arena would be part of the `Action` struct, holding the strings as well as the slice. Then in `Action.deinit()` just call `Arena.deinit()` for cleanup.

Reply:
That make a lot of sense, thank you.

I'll get to that next week.

Reply:
Added the `ArenaAllocator` to `PointerMapping`.

This is a really clean pattern. It can applied to `Mapping` and friends.